### PR TITLE
haystack/datadog context propagation

### PR DESCRIPTION
**Pull request by richardstartin**
This resolves the merge conflict in #1691 but otherwise uses the same baggage-based context propagation mechanism, and maintains the DD/haystack id conversion.

/cc @aantono

### CODE REVIEW

**Comment by richardstartin** [comment]
Won't merge until @aantono has given this a test drive. Will have to be a 0.60.0 feature.

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/HaystackHttpCodec.java [hunk 1]
```diff
@@ -139,9 +199,52 @@ public boolean accept(String key, String value) {
...
+      baggage.put(key, HttpCodec.decode(value));
+    }
+
     @Override
     protected int defaultSamplingPriority() {
       return PrioritySampling.SAMPLER_KEEP;
     }
   }
+
+  private static String convertBigIntToUUID(DDId id) {
```
Comment (aantono) [inline dd-trace-core/src/main/java/datadog/trace/core/propagation/HaystackHttpCodec.java:215]: Given this no longer deals with BigInteger, should we rename to `convertDDIdToUUID`?

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/HaystackHttpCodec.java [hunk 2]
```diff
@@ -139,9 +199,52 @@ public boolean accept(String key, String value) {
...
+    // This is not a true/real UUID, as we don't care about the version and variant markers
+    //  the creation is just taking the least significant bits and doing static most significant
+    // ones.
+    //  this is done for the purpose of being able to maintain cardinality and idempotence of the
+    // conversion
+    String idHex = String.format("%016x", id.toLong());
+    return DATADOG + "-" + idHex.substring(0, 4) + "-" + idHex.substring(4);
+  }
+
+  private static DDId convertUUIDToBigInt(String value) {
```
Comment (aantono) [inline dd-trace-core/src/main/java/datadog/trace/core/propagation/HaystackHttpCodec.java:225]: Given this no longer deals with BigInteger, should we rename to `convertUUIDToDDId`?

File: dd-trace-core/src/main/java/datadog/trace/core/propagation/HaystackHttpCodec.java [hunk 3]
```diff
@@ -32,14 +44,54 @@ private HaystackHttpCodec() {
...
+            getBaggageItemIgnoreCase(context.getBaggageItems(), HAYSTACK_TRACE_ID_BAGGAGE_KEY);
+        String injectedTraceId;
+        if (originalHaystackTraceId != null
+            && convertUUIDToBigInt(originalHaystackTraceId).equals(context.getTraceId())) {
+          injectedTraceId = originalHaystackTraceId;
+        } else {
+          injectedTraceId = convertBigIntToUUID(context.getTraceId());
+        }
+        setter.set(carrier, TRACE_ID_KEY, injectedTraceId);
+        context.setTag(HAYSTACK_TRACE_ID_BAGGAGE_KEY, injectedTraceId);
```
Comment (aantono) [inline dd-trace-core/src/main/java/datadog/trace/core/propagation/HaystackHttpCodec.java:66]: Is there a particular appropriate convention for tag names that’s used within DD ecosystem, to make sure the tag names are compliant here and in other places?

(Also, do we want to add the same tags of original ids in Zipkin codec as well?)

**Comment by richardstartin** [comment]
> Overall the code looks correct! I will make a test build tomorrow to try the agent locally just to double check the headers propagation, etc.

@aantono any luck?

**Comment by aantono** [comment]
@richardstartin Looks good on my local machine tests...

**Pull request merged**